examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+    ) -> None:
+
+        self.fabric = Fabric(
+            accelerator=accelerator,
+            strategy=strategy,
+            devices=devices,
+            precision=precision,
+            plugins=plugins,
+            callbacks=callbacks,
+            loggers=loggers,
+        )
+        self.global_step = 0
+        self.grad_accum_steps: int = grad_accum_steps
+        self.current_epoch = 0
+
+        self.max_epochs = max_epochs
+        self.max_steps = max_steps
+        self.should_stop = False
+
+        # ensures limit_X_batches is either int or inf
+        if not isinstance(limit_train_batches, int):
+            assert limit_train_batches == float("inf")
+
+        if not isinstance(limit_val_batches, int):
+            assert limit_val_batches == float("inf")
+
+        self.limit_train_batches = limit_train_batches
+        self.limit_val_batches = limit_val_batches
+        self.val_sanity_check = val_sanity_check
+        self.validation_frequency = validation_frequency
+        self.use_distributed_sampler = use_distributed_sampler
+        self._current_train_return: Union[torch.Tensor, Mapping[str, Any]] = {}
+        self._current_val_return: Optional[Union[torch.Tensor, Mapping[str, Any]]] = {}
+
+        self.checkpoint_dir = checkpoint_dir
+        self.checkpoint_frequency = checkpoint_frequency
+
+    def fit(
+        self, model: LightningModule, train_loader: DataLoader, val_loader: DataLoader, ckpt_path: Optional[str] = None
+    ):
+        if self.fabric.is_global_zero:
+            self.fabric.call("prepare_data")
+
+        # TODO: Have fabric launch the loop function directly?
+        self.fabric.launch()
+
+        # setup dataloaders
+        train_loader = self.fabric.setup_dataloaders(train_loader, use_distributed_sampler=self.use_distributed_sampler)
+        if val_loader is not None:
+            val_loader = self.fabric.setup_dataloaders(val_loader, use_distributed_sampler=self.use_distributed_sampler)
+
+        # setup model and optimizer
+        if isinstance(self.fabric.strategy, FSDPStrategy):
+            model = self.fabric.setup_module(model)
+            # FIXME: currently, there is no way to support fsdp with model.configure_optimizers in fabric
+            # as it would require fabric to hold a reference to the model, which we don't want to.
+            raise NotImplementedError("BYOT currently does not support FSDP")
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            optimizer = self.fabric.setup_optimizers(optimizer)
+        else:
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            assert optimizer is not None
+            model, optimizer = self.fabric.setup(model, optimizer)
+
+        # assemble state (current epoch and global step will be added in save)
+        state = {"model": model, "optim": optimizer, "scheduler": scheduler_cfg}
```

Comment:
Do you see a problem including them here as initial state directly?

Reply:
Yes there is. Python has references for everything, except for integers and a few characters, which are handled with call by value, meaning that it would  copy the value of `current_epoch` and `global_step` if added here instead of just adding a reference. 

This would result in the issue that later changes to these variables would not be reflected in the state dict.